### DISCUSSION TITLE
ZFS caching in action on SMB Direct with 100gbit connection... but locally crashes.

### DISCUSSION THREAD

**Discussion by besterino** [openzfsonwindows/openzfs#416] [Show and tell]
So, ZFS delivers more than twice the physical speed of the underlying NVME SSD over a network connection. Impressive!

![image](https://github.com/user-attachments/assets/8848043e-d118-4fe3-8bd0-3ada2e9be01e)

Unfortunately, however, running chrystal disk mark locally on the ZFS pool crashed the whole server... after reboot the pool was lost (zpool status --> no pools available). 

More details: 

Windows Server 2025 Desktop Experience (evaluation) 
ZFSonWindows 2.2.6-rc9
AMD Threadripper 3960X
64GB RAM
Various SSDs 

Created ZFS pool on clean 2TB NVME SSD with: "zpool create -O casesensitivity=insensitive -o ashift=12 -O atime=off zfswin physicaldrive2"

Set driveletter with: zfs set driveletter=P zfswin

Running CDM on P: hangs for long time at "Preparing...Sequential Read". Clicking on "Stop" in CDM does nothing. After a while Server crashed and rebooted. Event log: 

> Faulting application name: zfs.exe, version: 1.0.0.1, time stamp: 0x6733ebcb
> Faulting module name: ntdll.dll, version: 10.0.26100.2033, time stamp: 0x9b03969e
> Exception code: 0xc0000005
> Fault offset: 0x000000000001a9dd
> Faulting process id: 0x11EC
> Faulting application start time: 0x1DB3912D569C588
> Faulting application path: C:\Program Files\OpenZFS On Windows\zfs.exe
> Faulting module path: C:\WINDOWS\SYSTEM32\ntdll.dll
> Report Id: 40b74776-ddda-46c4-9045-93fe816a032e
> Faulting package full name: 
> Faulting package-relative application ID: 

When trying to reproduce the behaviour, initially only CDM hung and again I could not stop the benchmark. But Server crashed again when I tried to terminate CDM through taskmanager when clicking on "Stop" in CDM did nothing. At least after second crash pool was still available and able for import.

Simply writing/reading (file copy) on the ZFS pool seems to work.

CDM on same hardware / SSD with NTFS volume runs without errors. 

Has anyone else experienced problems with crystal disk mark on a ZFS pool?

**Comment by lundman** [comment]
We had an issue with Crystal Disk earlier in the year, but I believed it was fixed. Either way, I'll have a look at it again.

**Comment by lundman** [comment]
I get deadlock alert running CrystalMark, so must be leaking a lock somewhere, I'll dig deeper.

**Comment by lundman** [comment]
OK so we call

```
			Status = fs_write_impl(DeviceObject, Irp, IrpSp,
			    wait, FALSE);
		}

	} except(EXCEPTION_EXECUTE_HANDLER) {
		Status = GetExceptionCode();
	}
```

and inside `fs_write_impl()` we get to around here:

```
	Status = zfs_write(zp, &uio, 0, NULL);

	if (!locked)
		MmUnlockPages(Irp->MdlAddress);

	if (!NT_SUCCESS(Status)) {
		dprintf("zfs_write returned %08lx\n", Status);
		goto end;
	}
```

in that `zfs_write()` completes and returns 0, then 2/3 steps below that we trigger an exception, which brings us all the way back to the `except`. At this point we have not unlocked the locks either.

We should;
1) unlock the locks when exceptions happen
2) work out why we get an exception, most likely due to UserPtr not mapped.

**Reply by lundman** [reply]
OK I have fixed this. But a mount issue came up, so let me address that as well before rc10.

**Comment by lundman** [comment]
OK I have pushed out rc10 which I hope will make CrystalDisk run better.

**Comment by besterino** [comment]
Thank you! Will test soon.
